src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncMainTests.cs

```diff
@@ -0,0 +1,827 @@
+                //     static Task Main() {
+                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "Task").WithArguments("System.Threading.Tasks.Task", "GetAwaiter").WithLocation(6, 12),
+                // (6,17): warning CS0028: 'Program.Main()' has the wrong signature to be an entry point
+                //     static Task Main() {
+                Diagnostic(ErrorCode.WRN_InvalidMainSig, "Main").WithArguments("Program.Main()").WithLocation(6, 17),
+                // error CS5001: Program does not contain a static 'Main' method suitable for an entry point
+                Diagnostic(ErrorCode.ERR_NoEntryPoint));
+        }
+
+        [Fact]
+        public void AsyncEmitMainOfIntTest()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+
+class Program {
+    static async Task<int> Main() {
+        Console.Write(""hello "");
+        await Task.Factory.StartNew(() => 5);
+        Console.Write(""async main"");
+        return 10;
+    }
+}";
+            var c = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_1));
+            var verifier = CompileAndVerify(c, expectedOutput: "hello async main", expectedReturnCode: 10);
+
+        }
+
+        [Fact]
+        public void AsyncEmitMainTest()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+
+class Program {
+    static async Task Main() {
+        Console.Write(""hello "");
+        await Task.Factory.StartNew(() => 5);
+        Console.Write(""async main"");
+    }
+}";
+            var c = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_1));
+            var verifier = CompileAndVerify(c, expectedOutput: "hello async main", expectedReturnCode: 0);
+        }
+
+        [Fact]
+        public void AsyncEmitMainTestCodegenWithErrors()
+        {
+            var source = @"
+using System;
+using System.Threading.Tasks;
+
+class Program {
+    static async Task<int> Main() {
+        Console.WriteLine(""hello"");
+        await Task.Factory.StartNew(() => 5);
+        Console.WriteLine(""async main"");
+    }
+}";
+            var c = CreateCompilationWithMscorlib45(source, options: TestOptions.DebugExe, parseOptions: TestOptions.Regular.WithLanguageVersion(LanguageVersion.CSharp7_1));
+            c.VerifyEmitDiagnostics(
+                // (6,28): error CS0161: 'Program.Main()': not all code paths return a value
+                //     static async Task<int> Main() {
+                Diagnostic(ErrorCode.ERR_ReturnExpected, "Main").WithArguments("Program.Main()").WithLocation(6, 28));
```

Comment:
Are these emit/codegen diagnostics? If not, consider changing the name of this method and `_StringArgs` version later.